File: water-loggers/water_tracker.lua
local waterTracker = {}
waterTracker.totalIntake = 0

local MENU_OPTIONS = {
    LOG_INTAKE = "1",
    VIEW_INTAKE = "2",
    RESET_INTAKE = "3",
    EXIT = "4"
}

function waterTracker.logIntake(amount)
    if amount > 0 then
        waterTracker.totalIntake = waterTracker.totalIntake + amount
        print("Logged " .. amount .. "ml of water.")
    else
        print("Please enter a valid amount.")
    end
end

function waterTracker.viewIntake()
    print("Total water intake for today: " .. waterTracker.totalIntake .. "ml")
end

function waterTracker.resetIntake()
    waterTracker.totalIntake = 0
    print("Water intake reset for the day.")
end

function waterTracker.showMenu()
    print("\nWater Tracker Menu:")
    print("1. Log water intake")
    print("2. View total intake")
    print("3. Reset water intake")
    print("4. Exit")
end

function waterTracker.getUserInput(prompt)
    print(prompt)
    return io.read()
end

while true do
    waterTracker.showMenu()
    local choice = waterTracker.getUserInput("Please select an option:")

    if choice == MENU_OPTIONS.LOG_INTAKE then
        local amount = tonumber(waterTracker.getUserInput("Enter amount of water in ml:"))
        if amount then
            waterTracker.logIntake(amount)
        else
            print("Invalid input. Please enter a numeric value.")
        end
    elseif choice == MENU_OPTIONS.VIEW_INTAKE then
        waterTracker.viewIntake()
    elseif choice == MENU_OPTIONS.RESET_INTAKE then
        waterTracker.resetIntake()
    elseif choice == MENU_OPTIONS.EXIT then
        print("Exiting the water tracker app.")
        break
    else
        print("Invalid choice. Please try again.")
    end
end

File: water-loggers/water_tracker.scala
import scala.io.StdIn._

object WaterTracker {
  var totalWaterIntake: Double = 0.0
  val logOption = "1"
  val viewOption = "2"
  val exitOption = "3"

  def main(args: Array[String]): Unit = {
    println("Welcome to the Water Tracker App!")
    var running = true

    while (running) {
      displayMenu()
      val choice = readLine()

      choice match {
        case `logOption` => logWaterIntake()
        case `viewOption` => viewTotalWaterIntake()
        case `exitOption` => 
          running = false
          println("Exiting the Water Tracker App. Stay hydrated!")
        case _ => println("Invalid option. Please try again.")
      }
    }
  }

  def displayMenu(): Unit = {
    println("\nPlease choose an option:")
    println("1. Log water intake")
    println("2. View total water intake")
    println("3. Exit")
  }

  def logWaterIntake(): Unit = {
    println("Enter the amount of water (in liters) you drank:")
    try {
      val amount = readLine().toDouble
      if (amount < 0) {
        println("Please enter a positive number.")
      } else {
        totalWaterIntake += amount
        println(s"You have logged $amount liters of water.")
        println(s"Total water intake is now $totalWaterIntake liters.")
      }
    } catch {
      case _: NumberFormatException => println("Invalid input. Please enter a numeric value.")
    }
  }

  def viewTotalWaterIntake(): Unit = {
    println(f"Your total water intake today is $totalWaterIntake%.2f liters.")
  }
}
